### ISSUE TITLE
Stale information drawn by nbAgg

### ISSUE THREAD

**Issue by ianhi** [matplotlib/matplotlib#19116]
### Bug report
This is a follow up to the glitching noted in https://github.com/matplotlib/matplotlib/pull/19059.

**Bug summary**
If you are not forcing a full frontend redraw and are sending messages to the frontend fast enough then there can be artifacts left in the plot. 

**Code for reproduction**
```python
import itertools

cnt = itertools.count()
bg = None

def onclick_handle(event):
    """Should draw elevating green line on each mouse click"""
    global bg
    if bg is None:
        bg = ax.figure.canvas.copy_from_bbox(ax.bbox)        
    ax.figure.canvas.restore_region(bg)

    cur_y = (next(cnt) % 10) * 0.1
    ln.set_ydata([cur_y, cur_y])
    ax.draw_artist(ln)
    ax.figure.canvas.blit(ax.bbox)


fig, ax = plt.subplots()
ax.plot([0, 1], [0, 1], 'r')
ln, = ax.plot([0, 1], [0, 0], 'g', animated=True)
plt.show()
ax.figure.canvas.draw()

ax.figure.canvas.mpl_connect('button_press_event', onclick_handle)
```

**Actual outcome**
![smearing](https://user-images.githubusercontent.com/10111092/100963023-910a4a00-34f3-11eb-80a3-97668948dbe2.gif)

**Expected outcome**
Only one green line should be ever be present.

**Matplotlib version**

  * Operating system:
  * Matplotlib version: Master
  * Matplotlib backend (`print(matplotlib.get_backend())`): nbAgg or ipympl
  * Python version: 
  * Jupyter version (if applicable): Jupyterlab 2.2.9

In https://github.com/matplotlib/matplotlib/pull/19059#issuecomment-737707430 @tacaswell  noted the following:

> @danielballan has also noticed this form of glitching in ipympl and is interested in fixing it. I agree the issue is that via some mechanism one or more of the diff frames are being lost or processed out of order. We either need to make the front ends smart enough to detect this (by tacking a frame number on to everything and verifying that that they come in order) or on the kernel side force a non-diff every N frame (aka key-frames analogous to the way that mpeg works internally).

**Comment by ianhi** [comment]
Wasn't sure on the best title for this issue, open to suggestions

**Comment by story645** [comment]
Does it only happen w/ the jupyter backends or is it a general animation glitch?

**Comment by ianhi** [comment]
I can't to replicate it with the tkAgg backend so I think it is specific the juptyer backends

**Comment by ianhi** [comment]
Actually, it may also occur if you embed webAgg into a generic webpage. But I also may have found a bug with this example https://matplotlib.org/3.3.1/gallery/user_interfaces/embedding_webagg_sgskip.html. It seems I need to refresh in order to have updates the graph actually get updated. See this GIF:
![Peek 2020-12-16 00-58](https://user-images.githubusercontent.com/10111092/102310881-dab65400-3f39-11eb-9955-3dbb5ee741c7.gif)
<details>
 <summary>Embed Code</summary>

```python

import io
import json
import mimetypes
from pathlib import Path

try:
    import tornado
except ImportError as err:
    raise RuntimeError("This example requires tornado.") from err
import tornado.web
import tornado.httpserver
import tornado.ioloop
import tornado.websocket


import matplotlib as mpl
from matplotlib.backends.backend_webagg_core import (
    FigureManagerWebAgg, new_figure_manager_given_figure)
from matplotlib.figure import Figure

import numpy as np


def create_figure():
    """
    Creates a simple example figure.
    """

    return fig


# The following is the content of the web page.  You would normally
# generate this using some sort of template facility in your web
# framework, but here we just use Python string formatting.
html_content = """
<html>
  <head>
    <!-- TODO: There should be a way to include all of the required javascript
               and CSS so matplotlib can add to the set in the future if it
               needs to. -->
    <link rel="stylesheet" href="_static/css/page.css" type="text/css">
    <link rel="stylesheet" href="_static/css/boilerplate.css"
          type="text/css" />
    <link rel="stylesheet" href="_static/css/fbm.css" type="text/css" />
    <link rel="stylesheet" href="_static/css/mpl.css" type="text/css">
    <script src="mpl.js"></script>

    <script>
      /* This is a callback that is called when the user saves
         (downloads) a file.  Its purpose is really to map from a
         figure and file format to a url in the application. */
      function ondownload(figure, format) {
        window.open('download.' + format, '_blank');
      };

      function ready(fn) {
        if (document.readyState != "loading") {
          fn();
        } else {
          document.addEventListener("DOMContentLoaded", fn);
        }
      }

      ready(
        function() {
          /* It is up to the application to provide a websocket that the figure
             will use to communicate to the server.  This websocket object can
             also be a "fake" websocket that underneath multiplexes messages
             from multiple figures, if necessary. */
          var websocket_type = mpl.get_websocket_type();
          var websocket = new websocket_type("%(ws_uri)sws");

          // mpl.figure creates a new figure on the webpage.
          var fig = new mpl.figure(
              // A unique numeric identifier for the figure
              %(fig_id)s,
              // A websocket object (or something that behaves like one)
              websocket,
              // A function called when a file type is selected for download
              ondownload,
              // The HTML element in which to place the figure
              document.getElementById("figure"));
        }
      );
    </script>

    <title>matplotlib</title>
  </head>

  <body>
    <div id="figure">
    </div>
  </body>
</html>
"""


class MyApplication(tornado.web.Application):
    class MainPage(tornado.web.RequestHandler):
        """
        Serves the main HTML page.
        """

        def get(self):
            manager = self.application.manager
            ws_uri = "ws://{req.host}/".format(req=self.request)
            content = html_content % {
                "ws_uri": ws_uri, "fig_id": manager.num}
            self.write(content)

    class MplJs(tornado.web.RequestHandler):
        """
        Serves the generated matplotlib javascript file.  The content
        is dynamically generated based on which toolbar functions the
        user has defined.  Call `FigureManagerWebAgg` to get its
        content.
        """

        def get(self):
            self.set_header('Content-Type', 'application/javascript')
            js_content = FigureManagerWebAgg.get_javascript()

            self.write(js_content)

    class Download(tornado.web.RequestHandler):
        """
        Handles downloading of the figure in various file formats.
        """

        def get(self, fmt):
            manager = self.application.manager
            self.set_header(
                'Content-Type', mimetypes.types_map.get(fmt, 'binary'))
            buff = io.BytesIO()
            manager.canvas.figure.savefig(buff, format=fmt)
            self.write(buff.getvalue())

    class WebSocket(tornado.websocket.WebSocketHandler):
        """
        A websocket for interactive communication between the plot in
        the browser and the server.

        In addition to the methods required by tornado, it is required to
        have two callback methods:

            - ``send_json(json_content)`` is called by matplotlib when
              it needs to send json to the browser.  `json_content` is
              a JSON tree (Python dictionary), and it is the responsibility
              of this implementation to encode it as a string to send over
              the socket.

            - ``send_binary(blob)`` is called to send binary image data
              to the browser.
        """
        supports_binary = True

        def open(self):
            # Register the websocket with the FigureManager.
            manager = self.application.manager
            manager.add_web_socket(self)
            if hasattr(self, 'set_nodelay'):
                self.set_nodelay(True)

        def on_close(self):
            # When the socket is closed, deregister the websocket with
            # the FigureManager.
            manager = self.application.manager
            manager.remove_web_socket(self)

        def on_message(self, message):
            # The 'supports_binary' message is relevant to the
            # websocket itself.  The other messages get passed along
            # to matplotlib as-is.

            # Every message has a "type" and a "figure_id".
            message = json.loads(message)
            if message['type'] == 'supports_binary':
                self.supports_binary = message['value']
            else:
                manager = self.application.manager
                manager.handle_json(message)

        def send_json(self, content):
            self.write_message(json.dumps(content))

        def send_binary(self, blob):
            if self.supports_binary:
                self.write_message(blob, binary=True)
            else:
                data_uri = "data:image/png;base64,{0}".format(
                    blob.encode('base64').replace('\n', ''))
                self.write_message(data_uri)

    def __init__(self, figure):
        self.figure = figure
        self.manager = new_figure_manager_given_figure(id(figure), figure)

        super().__init__([
            # Static files for the CSS and JS
            (r'/_static/(.*)',
             tornado.web.StaticFileHandler,
             {'path': FigureManagerWebAgg.get_static_file_path()}),

            # Static images for the toolbar
            (r'/_images/(.*)',
             tornado.web.StaticFileHandler,
             {'path': Path(mpl.get_data_path(), 'images')}),

            # The page that contains all of the pieces
            ('/', self.MainPage),

            ('/mpl.js', self.MplJs),

            # Sends images and events to the browser, and receives
            # events from the browser
            ('/ws', self.WebSocket),

            # Handles the downloading (i.e., saving) of static images
            (r'/download.([a-z0-9.]+)', self.Download),
        ])


import itertools

if __name__ == "__main__":
    cnt = itertools.count()
    fig = Figure()
    application = MyApplication(fig)
    ax = fig.add_subplot(111)
    ax.plot([0, 1], [0, 1], 'r')
    ln, = ax.plot([0, 1], [0, 0], 'g', animated=False)
    ax.figure.canvas.draw()
    fig.canvas.draw()

    bg = ax.figure.canvas.copy_from_bbox(ax.bbox)
    times_clicked = 0

    def onclick_handle(event):
        """Should draw elevating green line on each mouse click"""
        global times_clicked
        times_clicked += 1
        ax.figure.canvas.restore_region(bg)
        print('weeeeeeeeeee')

        cur_y = (next(cnt) % 10) * 0.1
        ln.set_ydata([cur_y, cur_y])
        ax.draw_artist(ln)
        ax.figure.canvas.blit(ax.bbox)
        ax.set_title(f"clicked: {times_clicked} times")

    ax.figure.canvas.mpl_connect('button_press_event', onclick_handle)

    http_server = tornado.httpserver.HTTPServer(application)
    http_server.listen(8080)

    print("http://127.0.0.1:8080/")
    print("Press Ctrl+C to quit")

    tornado.ioloop.IOLoop.instance().start()
```
</details>
but I think that's a separate issue that deserves to be filed in own right

**Comment by tacaswell** [comment]
Can you try wrapping the callback in a threading lock?  I am worried that somehow threads are getting involved (I recently skimmed the ipykernel code and I _think_ it is single threaded, but given that zmq and zmq calbacks are involved I am not 100% sure of that).  A way to get this effect without errors in the sending is

  1. thread A: restore
  2. thread B: restore
  3. thread A: draw
  4. thread B: draw

However it is not clear to me how that survives in the double buffer on the Python side I am not clear (as even if you ended up with the same wrong version in both buffers, on the next slow click through I think it should be cleaned up.

It would also be interesting to see if you make the axes background 50% green if you can see a "negative line" anyplace.   Maybe make the axes background (128, 128, 128) and then randomize the line color?

I'm still mostly sure that this is a dropped diff issue, but it might be worth being very sure of that.

----

What version of mpl and browser? @QuLogic has been making progress on cleaning up the js and either you may not have a fix or we put a regression in.    

What are those dom exceptions referring to?

**Comment by ianhi** [comment]
> Can you try wrapping the callback in a threading lock?

I will poke around with this sometime in the next few days.

> It would also be interesting to see if you make the axes background 50% green if you can see a "negative line" anyplace. Maybe make the axes background (128, 128, 128) and then randomize the line color?

This is a nice idea, I will try this out.


> What version of mpl and browser?

Firefox 83.0 on Ubuntu, matplotlib: 3.3.1

DOM exceptions (not even a screenshot! I guess I'm learning)

```
Uncaught DOMException: An attempt was made to use an object that is not, or is no longer, usable mpl.js:377
    send_message http://127.0.0.1:8080/mpl.js:377
    request_resize http://127.0.0.1:8080/mpl.js:371
    resizeObserver http://127.0.0.1:8080/mpl.js:218
    (Async: ResizeObserverCallback)
    _init_canvas http://127.0.0.1:8080/mpl.js:169
    figure http://127.0.0.1:8080/mpl.js:55
    <anonymous> http://127.0.0.1:8080/:40
    (Async: EventListener.handleEvent)
    ready http://127.0.0.1:8080/:26
    <anonymous> http://127.0.0.1:8080/:30
```

**Comment by ianhi** [comment]
I tried the webAgg example on master matplotlib and found it's broken there as well. So I opened https://github.com/matplotlib/matplotlib/issues/19129 to keep it separate from this issue

**Comment by ianhi** [comment]
This same glitching does **not** occur when using the webAgg backend (tested when running the branch for https://github.com/matplotlib/matplotlib/pull/19131) so I'd guess it's an issue either with threads in ipython, or with the lagginess of the jupyter comms. 

TBD on testing with a threading lock, but I have some thoughts on the latter. 

Pure websockets are significantly faster than the jupyter comms - see my investigation here: https://github.com/ianhi/widget_message_speed#widget_message_speed (there are matplotlib plots in the readme!). I don't really understand how those comms work, but maybe they're not actually a setting up a proper queue, and so it can end up out of order especially given the times that are in play?


Is it possible for maptlotlib to hijack the tornado event loop that underlies jupyter, add a new websocket handler, and then open a direct websocket to the frontend? This would bypass a bit how I think widgets are meant to work, but I think would also come with significant performance improvements for interactivity.

**Comment by QuLogic** [comment]
Initially, I thought the problem might have been #18583, and that the binary comm might be a separate channel, with things arriving in parallel, but not necessarily in order. However, reverting that change does _not_ fix the artifacting, though it maybe happens a bit less often (I'm not sure exactly how you tested other than clicking a lot.)

**Comment by QuLogic** [comment]
So the way display works is a bit funny; I think there may be a few extra buffers and async than necessary. We draw on a `Canvas`, which mostly makes sense. There is an 'image mode', which is either 'diff' or 'full', which determines whether the `Canvas` is cleared first. This mode is sent before the image data in a separate message. While this could be causing a mismatch, I think it might not be the problem.

When the JS side receives image data, it doesn't draw that directly on the `Canvas`. Instead, it creates a new object URL, and sets the source on an `Image`. Then in the `Image`'s `onload`, it does the actual draw on the `Canvas`. Now I believe this may be the correct way to do things due to what sources `Canvas` supports, but since we re-use the same `Image` object, I think there may be some room for the `Image` to trigger `onload` and start drawing while the websocket has received a new diff and started modifying the `Image` again.

I've been thinking maybe we should use `Canvas.putImageData`, but I think that would be raw bytes, and mean dropping the whole 'diff' image mode. Not encoding as PNG will save CPU, but increase transfer size. I haven't had a chance to try this out, and it may end up being slower because of that.

**Comment by QuLogic** [comment]
Some quick print debug shows that when I get an artifact, I see that 2-3 image data loads have happened, but only one draw in the `onload` function. Though I'm not sure how that created a double-draw of things.

Switching from a single `Image` to just making a new one every time _seems_ like it might have fixed it, but I don't know if I've clicked fast enough. It also might be leaking the old objects right now.

**Comment by tacaswell** [comment]
>  Not encoding as PNG will save CPU, but increase transfer size. 

I think Mike did a systematic study of this and the png compression time was worth it for reduced data (and CPUs have only gotten comparatively faster).

**Comment by ianhi** [comment]
> I've been thinking maybe we should use Canvas.putImageData, but I think that would be raw bytes, and mean dropping the whole 'diff' image mode. Not encoding as PNG will save CPU, but increase transfer size. I haven't had a chance to try this out, and it may end up being slower because of that.

What about generic compression of the bytes/array? For example ipydatawidgets provides serializers for numpy arrays to scijs arrays in the frontend (https://github.com/vidartf/ipydatawidgets#arrays) I discussed this some w.r.t ipycanvas https://github.com/martinRenou/ipycanvas/issues/104. The actual compressed serialization happens here: https://github.com/vidartf/ipydatawidgets/blob/da6db7a9f9ace74c87583e633b22c3634526520e/ipydatawidgets/ndarray/serializers.py#L94 and on the js side you can use pako to decompress: https://github.com/vidartf/ipydatawidgets/blob/da6db7a9f9ace74c87583e633b22c3634526520e/packages/serializers/src/compression.ts#L16



> but since we re-use the same Image object, I think there may be some room for the Image to trigger onload and start drawing while the websocket has received a new diff and started modifying the Image again.

I agree that this seems the most likely culprit. Maybe some sort of lock or usage of `await` somewhere could help with this?

**Comment by QuLogic** [comment]
> > but since we re-use the same Image object, I think there may be some room for the Image to trigger onload and start drawing while the websocket has received a new diff and started modifying the Image again.
> 
> I agree that this seems the most likely culprit. Maybe some sort of lock or usage of `await` somewhere could help with this?

This seemed possibly okay, and didn't leak (kind of depends on JS garbage collection, I think), but it is difficult to test. I'm trying to put together something in selenium, but only started with WebAgg for now.

**Comment by ianhi** [comment]
I think the thing may be to do:

```
await image.decode()
```
immediately after setting the src.

**Comment by tacaswell** [comment]
@ianhi To verify, blitting is still broken with webagg/nbagg/ipympl?

**Comment by ianhi** [comment]
On the current matplotlib master it was not happening in `webagg` but was for `nbagg`. I also tried with ipympl and was not able to replicate it, but in the past I trouble doing it consistently so I don't expect that that is fixed.

**Comment by raphaelquast** [comment]
hey, any updates on this? I still get this issue with `ipympl` and I have no idea how to even temporarily address this...

**Comment by raphaelquast** [comment]
Half a year later so I give it another try to make sure it stays on the radar 😄 

Is there anything I can do to help getting this fixed? 
Unfortunately I have limited javascript knowledge so I can't help on the server side... I looked into the webagg backend code to see if I can come up with some python-side fix but so far I had no luck...

As I see it, this issue is basically the only reason why interactive matplotlib plots are not usable in jupyter notebooks... and I think that would be extremely nice to have since they are nowadays used almost everywhere,,,.

**Comment by tacaswell** [comment]
To try and summarize:
 - to save network I/O we will send diffs rather than the full image every time
 - on the Python side we do this via a double-buffer scheme where we keep a copy of the _last_ render around to send it
 - if the Python server and js client ever get out of sync and drop a diff on the floor you will get this sort of artifact (the diff that _adds_ the line gets processed, the diff that removes it does not and the diff that puts it someplace else does -> 2 lines)

I think the best fix is to understand where the drops are coming from (I could imagine that the js side is debouncing us somehow and just not applying a few of the diffs or that there is (accidentally) some threading going on on the Python side and our double-buffer scheme is getting out of sync (I had some _very_ fun bugs around this with some of my BNL code)) and solve that problem. 

A practical solution on the Python side could be to crib off the "key frame" idea from mp4 compression.  A scheme were we send the full image every N draws or M seconds might be enough to paper over what ever the real bug is....

**Comment by ianhi** [comment]
> A practical solution on the Python side could be to crib off the "key frame" idea from mp4 compression. A scheme were we send the full image every N draws or M seconds might be enough to paper over what ever the real bug is....

building on this maybe we could send extra information with each diff that allows the js to check if it's in sync with the python, and if not then it requests a full image to reset itself

**Comment by raphaelquast** [comment]
OK, so I did some checking and while the "key-frame" method works in the sense that it removes remaining glitches from previous blit-events, it is in my opinion no usable solution because 
- it causes the screen to "flicker" whenever a full re-draw is triggered
- on more complex figures nearly every blit-event causes a glitch... so it would basically come down to "force_full=True"

  <details> 
  <summary><b>[click]</b> for a gif that shows severe glitching with a complex figure (in a jupyter-notebook using `ipympl` as backend) </summary>
  
  ![mpl_blit_issue](https://user-images.githubusercontent.com/22773387/204394682-6645b2b8-c031-4e65-814c-f8e93026a6dc.gif)
  
  </details>

While testing, i however noticed that within `ipympl` the `"ack"` message does not seem to trigger...
(it works just fine in `webagg` where also blitting works as expected)
- From the comment it seems that this message is primarily sent to reduce network-latency so maybe sending it in `ipympl` might also yield some improvement??


https://github.com/matplotlib/matplotlib/blob/041d4e02e60efef1df78d79e34ae0488d8b337ee/lib/matplotlib/backends/backend_webagg_core.py#L272-L279

**Comment by ianhi** [comment]
@raphaelquast can you post code for your example that reproducibly causes this issue? That would make it much easier to debug - my example has me mashing the mouse only to have it bug sometimes.

**Comment by raphaelquast** [comment]
@ianhi of course! 
I think it's enough to just make the changes faster in your initial example... 
Something like this already creates quite a lot of glitches in a jupyter-notebook:

```python
%matplotlib widget
import matplotlib.pyplot as plt
import numpy as np

def onclick_handle(event):
    global bg
    ax.figure.canvas.restore_region(bg)
    ln.set_ydata([event.ydata, event.ydata])
    ax.draw_artist(ln)
    ax.figure.canvas.blit(ax.bbox)

fig, ax = plt.subplots()
ax.plot([0, 1], [0, 1], 'r')
ln, = ax.plot([0, 1], [0, 0], 'g', animated=True)
plt.show()
ax.figure.canvas.draw()
bg = ax.figure.canvas.copy_from_bbox(ax.bbox)        

ax.figure.canvas.mpl_connect('motion_notify_event', onclick_handle)
```

**Comment by ianhi** [comment]
thanks that's helpful!

In the end I found that modifying the original example (wiht a set offset) but using mousemove events is the best for debugging. 
```python
%matplotlib widget
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.plot([0, 1], [0, 1], 'r')
ln, = ax.plot([0, 1], [0, 0], 'g', animated=True)
ax.figure.canvas.draw()
bg = fig.canvas.copy_from_bbox(ax.bbox)

import itertools

cnt = itertools.count()
def onclick_handle(event):
    global bg, ypos
    ax.figure.canvas.restore_region(bg)
    ln.set_ydata([event.ydata, event.ydata])
    
    cur_y = (next(cnt) % 10) * 0.1
    ln.set_ydata([cur_y, cur_y])
    ax.draw_artist(ln)
    ax.figure.canvas.blit(ax.bbox)

fig.canvas.mpl_connect('motion_notify_event', onclick_handle)
plt.show()
```
**Some notes:**

I made a branch of ipympl which keeps track of image numbers, and also logs the diffs that get sent: https://github.com/ianhi/ipympl/commit/0e3ea97822d08dad9298a38336ed043240056c61

Example console log:

```
Object { type: "binary", number: 0 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Error setting state: CanvasRenderingContext2D.drawImage: Passed-in image is "broken" [272.eb0d340d1eec40ac564f.js:1:22003](http://localhost:8889/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/272.eb0d340d1eec40ac564f.js)
Object { type: "binary", number: 1 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 2 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 3 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 4 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 5 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 6 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 7 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 8 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 9 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 10 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 11 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 12 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 13 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 14 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 15 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 16 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 17 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 18 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 19 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 20 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 21 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 22 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 23 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 24 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 25 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 26 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 27 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 28 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 29 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 30 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 31 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 32 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 33 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 34 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 35 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 36 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 37 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 38 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 39 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 40 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 41 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 42 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 43 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 44 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 45 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 46 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 47 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 48 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
Object { type: "binary", number: 49 }
[mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
```


so it seems as though the images are being received and dealt with in the order that they are sent.

You can then visualize each diff that was sent with this code:

```python
import io
import numpy as np
from PIL import Image
from mpl_interactions import indexer
from mpl_interactions import ipyplot as iplt

def get_img(data):
    return np.asarray(Image.open(io.BytesIO(data)))
imgs_sent = [get_img(b) for b in fig.canvas.datas[1:]]

debug_figure = plt.figure()
ctrls = iplt.imshow(indexer(imgs_sent), idx = np.arange(len(imgs_sent)))

# optionally save a gif
# ctrls.save_animation("out.gif", debug_figure, "idx", interval=200)
```

which also seems sensible:
![out](https://user-images.githubusercontent.com/10111092/208758161-42ba6393-802b-4954-8547-5414781f103f.gif)

**Comment by ianhi** [comment]
With this new change: https://github.com/ianhi/ipympl/commit/469787da6007c89d5f98b3ba8b633bdc93c27e29

i'm getting suspicious that when the messages get received too quickly by the frontend they start interefering with eachother. This leads to console logs with sections like this:

```
done draw [mpl_widget.js:237](webpack://jupyter-matplotlib/lib/mpl_widget.js)
done update [mpl_widget.js:241](webpack://jupyter-matplotlib/lib/mpl_widget.js)
93 [mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
93 [mpl_widget.js:167](webpack://jupyter-matplotlib/lib/mpl_widget.js)
93 [mpl_widget.js:172](webpack://jupyter-matplotlib/lib/mpl_widget.js)
93 [mpl_widget.js:174](webpack://jupyter-matplotlib/lib/mpl_widget.js)
done draw [mpl_widget.js:237](webpack://jupyter-matplotlib/lib/mpl_widget.js)
done update [mpl_widget.js:241](webpack://jupyter-matplotlib/lib/mpl_widget.js)
94 [mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
94 [mpl_widget.js:167](webpack://jupyter-matplotlib/lib/mpl_widget.js)
94 [mpl_widget.js:172](webpack://jupyter-matplotlib/lib/mpl_widget.js)
94 [mpl_widget.js:174](webpack://jupyter-matplotlib/lib/mpl_widget.js)
95 [mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
95 [mpl_widget.js:167](webpack://jupyter-matplotlib/lib/mpl_widget.js)
95 [mpl_widget.js:172](webpack://jupyter-matplotlib/lib/mpl_widget.js)
95 [mpl_widget.js:174](webpack://jupyter-matplotlib/lib/mpl_widget.js)
96 [mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
96 [mpl_widget.js:167](webpack://jupyter-matplotlib/lib/mpl_widget.js)
96 [mpl_widget.js:172](webpack://jupyter-matplotlib/lib/mpl_widget.js)
96 [mpl_widget.js:174](webpack://jupyter-matplotlib/lib/mpl_widget.js)
done draw [mpl_widget.js:237](webpack://jupyter-matplotlib/lib/mpl_widget.js)
done update [mpl_widget.js:241](webpack://jupyter-matplotlib/lib/mpl_widget.js)
97 [mpl_widget.js:162](webpack://jupyter-matplotlib/lib/mpl_widget.js)
97 [mpl_widget.js:167](webpack://jupyter-matplotlib/lib/mpl_widget.js)
97 [mpl_widget.js:172](webpack://jupyter-matplotlib/lib/mpl_widget.js)
```

**Comment by raphaelquast** [comment]
@ianhi I think I made some progress on this issue since I have finally found a way how to avoid the lag in `webagg` by accepting events only if image processing count is equal to the image send-count.... this works amazingly well and seems to solves all problems I had with blitting in `webagg` so far.

From what I know, I think this might also fix the issues in `ipympl` since there should no longer be a stack of pending images that get mixed up if processing is out of order. 

I put together a first proposal on how this might work for `webagg` here: #27125

The only problem is that the method relies on the return of an "ack" message once an image has been processed which is currently not returned by `ipympl`...

https://github.com/matplotlib/matplotlib/blob/1d1171fa04a7d1fc3c2826b593610a2f832374f0/lib/matplotlib/backends/web_backend/js/nbagg_mpl.js#L112)


Since I have little (or actually no) experience with javascript, I am unsure how I would achieve this in `ipympl`'s javascript code... 
Could you implement the `"ack"` message so we can check if this would also solve issues in Juypter Notebooks?

**Comment by AnyaPorter** [comment]
Have there been any updates on this issue?

**Comment by raphaelquast** [comment]
Hey, unfortunately no updates concerning existing jupyter-backends from my side since I can't fix the issues in the js-code... but in the meantime I've found a simple way how to get `webagg` based plots as widgets inside jupyter notebooks... 

Using the proposed event-accumulation fix for `webagg` from above (https://github.com/matplotlib/matplotlib/pull/27160), the resulting plot-widgets are fast, support blitting and don't show any glitches!

I started a discussion on how to implement this properly (with a first example how it might work) in the matplotlib discourse:  
https://discourse.matplotlib.org/t/blit-ready-jupyter-notebook-backend-based-on-webagg
